---
# **``Statistical Hypothesis Testing``**
### **``Practical Scenarios``**
---

## $$**Contents**$$
1. [Python Stats Libraries](#Python_Stats_Pkgs)
2. [One Population](#One_Population)
    1. [T Test or STUDENT-T or STUDENT Test](#Small_Samples)
    2. [Z Test](#Large_Samples)
    3. [Population Proportion](#POPULATION_PROPORTION)
    4. [Chi-Square Test](#CHI_SQUARE_TEST)
3. [Two Populations](#Two_Populations)
    1. [Large Independent Samples](#Large_Independent_Samples)
        1. [Pooled Large Independent Samples](#Pooled_Large_Independent_Samples)
        2. [Not-Pooled Large Independent Samples](#Not-Pooled_Large_Independent_Samples)
    2. [Small Independent Samples](#Small_Independent_Samples)
        1. [Pooled](#Pooled_Small_Independent_Samples)
        2. [Not Pooled](#Notpooled_Small_Independent_Samples)
    3. [Population Proportions](#Pop_proportions)
    4. [Dependent Samples](#Dependent_samples)
4. [ANOVA](#ANOVA_Testing)

## **``Python_Stats_Pkgs``**
---

Import handmade image here

##### **Python packages primarily used for statistical testing**

In [15]:
from statsmodels import api as sm

In [16]:
dir(sm.stats)

['CompareCox',
 'CompareJ',
 'CompareMeans',
 'DescrStatsW',
 'Describe',
 'FTestAnovaPower',
 'FTestPower',
 'FactoredPSDMatrix',
 'GofChisquarePower',
 'HetGoldfeldQuandt',
 'Mediation',
 'NormalIndPower',
 'NullDistribution',
 'RegressionFDR',
 'Runs',
 'SquareTable',
 'StratifiedTable',
 'TTestIndPower',
 'TTestPower',
 'Table',
 'Table2x2',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acorr_breusch_godfrey',
 'acorr_ljungbox',
 'anova_lm',
 'binom_test',
 'binom_test_reject_interval',
 'binom_tost',
 'binom_tost_reject_interval',
 'breaks_cusumolsresid',
 'breaks_hansen',
 'chisquare_effectsize',
 'cochrans_q',
 'compare_cox',
 'compare_j',
 'corr_clipped',
 'corr_nearest',
 'corr_nearest_factor',
 'corr_thresholded',
 'cov_cluster',
 'cov_cluster_2groups',
 'cov_hac',
 'cov_hc0',
 'cov_hc1',
 'cov_hc2',
 'cov_hc3',
 'cov_nearest',
 'cov_nearest_factor_homog',
 'cov_nw_panel',
 'cov_white_simple',
 'diagnostic'

In [17]:
from statsmodels.stats import weightstats

In [18]:
dir(weightstats)

['CompareMeans',
 'DescrStatsW',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 '_tconfint_generic',
 '_tstat_generic',
 '_zconfint_generic',
 '_zstat_generic',
 '_zstat_generic2',
 'cache_readonly',
 'np',
 'stats',
 'ttest_ind',
 'ttost_ind',
 'ttost_paired',
 'zconfint',
 'ztest',
 'ztost']

In [19]:
import scipy.stats as scipy_tests

In [20]:
dir(scipy_tests)

['PearsonRConstantInputWarning',
 'PearsonRNearConstantInputWarning',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_binned_statistic',
 '_constants',
 '_continuous_distns',
 '_discrete_distns',
 '_distn_infrastructure',
 '_distr_params',
 '_hypotests',
 '_multivariate',
 '_rvs_sampling',
 '_stats',
 '_stats_mstats_common',
 '_tukeylambda_stats',
 'absolute_import',
 'alpha',
 'anderson',
 'anderson_ksamp',
 'anglit',
 'ansari',
 'arcsine',
 'argus',
 'bartlett',
 'bayes_mvs',
 'bernoulli',
 'beta',
 'betaprime',
 'binned_statistic',
 'binned_statistic_2d',
 'binned_statistic_dd',
 'binom',
 'binom_test',
 'boltzmann',
 'boxcox',
 'boxcox_llf',
 'boxcox_normmax',
 'boxcox_normplot',
 'bradford',
 'brunnermunzel',
 'burr',
 'burr12',
 'cauchy',
 'chi',
 'chi2',
 'chi2_contingency',
 'chisquare',
 'circmean',
 'circstd',
 'circvar',
 'combine_pvalues',
 'contingency',
 'cosine',
 'crystalball',

##### **Other supporting packages**

In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

import statsmodels.api as sm
import statsmodels.stats.weightstats as s_tests
import scipy.stats as scipy_tests

---
# **``One_Population``**
---

Import hand made image here

## **Small_Samples**
#### ***T or STUDENT or STUDENT-T TESTING***
---

#### ***Question-1*** :: **A researcher thinks that average age of males getting married in Delhi is not more than 30. He samples 21 males and found that his claim is true. Test the claim at 95% level of confidence.**

In [22]:
sample_age = np.array([20,33,22,24,25,26,27,28,21,34,33,32,31,22,23,24,24,24,25,26,27])

In [23]:
len(sample_age)

21

In [24]:
sample_mean = sample_age.mean()

In [25]:
sample_mean

26.238095238095237

In [26]:
sample_std_dev = sample_age.var()

In [27]:
sample_std_dev

16.562358276643995

In [28]:
results = scipy_tests.ttest_1samp(sample_age,30)

In [29]:
results

Ttest_1sampResult(statistic=-4.133916563687664, pvalue=0.0005143638120022441)

In [30]:
results[0]

-4.133916563687664

In [31]:
results[1]

0.0005143638120022441

#### ***CASE - I :: With 95% Level of Confidence***

In [32]:
alpha = 0.05
if (results[0] < 0) & (results[1] <= alpha):
    print("Reject the Null Hypothesis at {} level of significance".format(alpha))
else:
    print("Fail to reject the Null Hypothesis at {} level of significance".format(alpha))

Reject the Null Hypothesis at 0.05 level of significance


##### ***Test Statistic :: Manual Calculation***

In [33]:
(sample_age.mean()-30)/(np.sqrt(sample_std_dev/len(sample_age)))

-4.236003949869238

In [34]:
(sample_age.mean()-30)/(np.sqrt(sample_std_dev/(len(sample_age)-1)))

-4.133916563687664

#### ***Difference in Test statistic value is because Python Statsmodels use DOF in the denominator instead of 'n' i.e. the number of records in sample.*** 


***Go to location ::*** c:\programdata\anaconda3\lib\site-packages\scipy\stats\


***Open python file ::*** stats.py

#### **Implementation in Statsmodels**

In [35]:
np.var(sample_age, 0, ddof=1)

17.390476190476193

In [36]:
np.var(sample_age, 0)

16.562358276643995

#### ***CASE - II :: With 99% Level of Confidence***

In [37]:
alpha = 0.01
if (results[0] < 0) & (results[1] <= alpha):
    print("Reject the Null Hypothesis at {} level of significance".format(alpha))
else:
    print("Fail to reject the Null Hypothesis at {} level of significance".format(alpha))

Reject the Null Hypothesis at 0.01 level of significance


#### ***Question-2*** :: **A researcher claims that average age of people buying their first car is greater than 26. He sampled ages of 24 people and found that his claim is true. Test the claim at 90% level of confidence.**

In [38]:
first_car_sample = [22,23,33,32,31,30,27,28,29,20,30,30,31,31,32,34,27,28,24,24,24,25,26,26]

In [39]:
len(first_car_sample)

24

In [40]:
first_car_sample_mean = round(np.mean(first_car_sample,dtype=float),3)

In [41]:
first_car_sample_mean

27.792

In [42]:
first_car_sample_stddev = round(np.std(first_car_sample,dtype=float,ddof=0),3)

In [43]:
first_car_sample_stddev

3.674

In [44]:
results = scipy_tests.ttest_1samp(first_car_sample,26)

In [45]:
results

Ttest_1sampResult(statistic=2.338741244971851, pvalue=0.02840406196767573)

#### ***CASE - I :: With 90% Level of Confidence***

In [46]:
loc = 0.90
alpha= 1-loc

if (results[0] > 0) & (results[1] < alpha):
    print("Researcher claim is right and he rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
else:
    print("Researcher claim is wrong and he fail to reject the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))

Researcher claim is right and he rejected the Null Hypothesis at 0.9 L.O.C and 0.1 L.O.S


#### ***CASE - II :: With 99% Level of Confidence***

In [47]:
loc = 0.99
alpha= 1-loc

if (results[0] > 0) & (results[1] < alpha):
    print("Researcher claim is right and he rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
else:
    print("Researcher claim is wrong and he fail to reject the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))

Researcher claim is wrong and he fail to reject the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


#### ***Question-3*** :: **A researcher claims that average height of people living in India is not 160 cms. He sampled heights of 25 people and found that his claim is true. Test the claim at different level of confidences 85%, 95% and 99%.**

In [48]:
avg_height_sample = np.array([100,140,160,150,145,130,130,150,150,165,165,164,160,165,180,182,182,184,183,177,178,176,189,193,158])

In [49]:
len(avg_height_sample)

25

In [50]:
avg_height_sample_mean = round(np.mean(avg_height_sample,dtype=float),3)

In [51]:
avg_height_sample_mean

162.24

In [52]:
avg_height_sample_variance = np.var(avg_height_sample,dtype=float,ddof=0)

In [53]:
avg_height_sample_variance

462.6624

In [54]:
avg_height_sample_stddev = round(np.sqrt(avg_height_sample_variance),3)

In [55]:
avg_height_sample_stddev

21.51

In [56]:
pop_mean = 160

test_stats_height = scipy_tests.ttest_1samp(avg_height_sample,pop_mean)

In [57]:
test_stats_height

Ttest_1sampResult(statistic=0.5101777757069099, pvalue=0.614587179842627)

In [58]:
def hyp_test(test_stats_result,loc,kind_of_test=['l','r','lr']):
    alpha = 1 - loc
    if (kind_of_test == 'l') & (test_stats_result[0] < 0) & (test_stats_result[1] < alpha):
        print("Researcher claim is right. Thus, rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
    elif (kind_of_test == 'r') & (test_stats_result[0] > 0) & (test_stats_result[1] < alpha):
        print("Researcher claim is right. Thus, rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
    elif (kind_of_test == 'lr') & (test_stats_result[1] < alpha):    
        print("Researcher claim is right. Thus, rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
    else:
        print("Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))

#### ***CASE - I :: With 85% Level of Confidence***

In [59]:
hyp_test(test_stats_height,0.85,'lr')

Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at 0.85 L.O.C and 0.15 L.O.S


#### ***CASE - II :: With 95% Level of Confidence***

In [60]:
hyp_test(test_stats_height,0.95,'lr')

Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at 0.95 L.O.C and 0.05 L.O.S


#### ***CASE - III :: With 99% Level of Confidence***

In [61]:
hyp_test(test_stats_height,0.99,'lr')

Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


## **Large_Samples**
#### ***Z TEST***
---

#### ***Question-4*** :: **A beer bottle factory produces atleast 30 cans a day. A newly appointed manager thinks that this number is not correct. He sampled cans production data of 40 random days. Test the claim at 95% level of confidences.**

In [62]:
beer_cans_sample = np.array([12,24,25,36,34,21,23,22,33,22,11,16,17,18,36,32,31,30,30,30,30,30,14,18,19,23,24,27,
                             28,29,26,26,26,23,32,10,9,22,28,29])

In [63]:
beer_cans_sample_mean = round(np.mean(beer_cans_sample,dtype='float'),3)
beer_cans_sample_stddev = round(np.std(beer_cans_sample,dtype='float',ddof=0),3)

In [64]:
beer_cans_sample_mean,beer_cans_sample_stddev

(24.4, 7.056)

In [65]:
beer_cans_results = s_tests.ztest(beer_cans_sample, value=30, alternative='smaller', ddof=0)

In [66]:
beer_cans_results

(-5.019343982920408, 2.592411755802454e-07)

In [67]:
hyp_test(beer_cans_results,0.95,'l')

Researcher claim is right. Thus, rejected the Null Hypothesis at 0.95 L.O.C and 0.05 L.O.S


#### ***Question-5*** :: **The government says that the average weight of males in Canada atleast 162.9 pounds. A researcher thinks this is too high. He sampled data of 39 canadian males. Test the claim at 99% level of confidences.**

In [68]:
canadian_male_sweight = np.array([120,240,250,360,340,210,230,220,330,220,110,160,170,180,360,320,310,300,300,300,300,300,140,
                                  180,190,230,240,270,280,290,260,260,260,230,320,100,90,280,290])

In [69]:
canadian_male_sweight_mean = np.mean(canadian_male_sweight,dtype='float')

In [70]:
canadian_male_sweight_stddev = np.std(canadian_male_sweight,dtype='float')

In [71]:
canadian_male_sweight_mean,canadian_male_sweight_stddev

(244.6153846153846, 71.35487886383511)

In [72]:
canadian_male_sweight_results = s_tests.ztest(canadian_male_sweight, value=162.9, alternative='smaller', ddof=0)

In [73]:
canadian_male_sweight_results

(7.151752220548922, 0.9999999999995716)

In [74]:
hyp_test(canadian_male_sweight_results,0.99,'l')

Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


#### ***Question-6*** :: **A study showed that on average women in a city had 1.48 kids. A researcher believes this number is incorrect. He surveyed 100 women in the city & finds that on average these women doesn't had 1.48 kids. Test the claim at 90% LOC.**

In [75]:
women_kids_sample = np.random.randint(low=1, high=4, size=100)

In [76]:
women_kids_sample

array([1, 1, 2, 2, 1, 1, 2, 2, 3, 2, 2, 3, 2, 2, 3, 3, 3, 2, 1, 3, 3, 3,
       3, 1, 2, 2, 2, 2, 1, 1, 3, 1, 1, 1, 2, 3, 1, 1, 2, 3, 1, 1, 1, 2,
       2, 1, 3, 3, 3, 3, 2, 2, 3, 1, 3, 1, 2, 2, 1, 1, 3, 3, 1, 3, 2, 2,
       3, 2, 1, 2, 2, 3, 1, 3, 1, 1, 3, 1, 3, 2, 1, 3, 1, 3, 2, 2, 1, 2,
       3, 2, 3, 2, 3, 1, 3, 2, 2, 1, 1, 3])

In [77]:
women_kids_sample_mean = np.mean(women_kids_sample,dtype='float')

In [78]:
women_kids_sample_mean

2.0

In [79]:
women_kids_sample_stddev = np.std(women_kids_sample,dtype='float',ddof=0)

In [80]:
women_kids_sample_stddev

0.812403840463596

In [81]:
women_kids_sample_results = s_tests.ztest(women_kids_sample, value=1.48, alternative='two-sided', ddof=0)

In [82]:
women_kids_sample_results

(6.4007575309253015, 1.5460793266910318e-10)

#### ***CASE - I :: With 90% Level of Confidence***

In [83]:
hyp_test(women_kids_sample_results,0.90,'lr')

Researcher claim is right. Thus, rejected the Null Hypothesis at 0.9 L.O.C and 0.1 L.O.S


#### ***CASE - II :: With 99% Level of Confidence***

In [84]:
hyp_test(women_kids_sample_results,0.99,'lr')

Researcher claim is right. Thus, rejected the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


#### ***CASE - III :: Changed the Sample Data***

In [85]:
women_kids_changed_sample = np.random.randint(low=0, high=4, size=100, dtype='int')

In [86]:
women_kids_changed_sample

array([3, 3, 0, 0, 1, 0, 3, 3, 1, 1, 0, 1, 3, 2, 0, 1, 3, 3, 1, 3, 2, 3,
       0, 2, 0, 2, 3, 0, 0, 2, 3, 0, 0, 2, 2, 1, 3, 0, 0, 3, 1, 2, 0, 3,
       2, 3, 1, 2, 1, 3, 3, 2, 2, 2, 0, 0, 1, 3, 1, 2, 0, 1, 1, 1, 1, 2,
       2, 3, 2, 3, 1, 1, 2, 1, 3, 0, 1, 2, 0, 2, 2, 1, 3, 1, 3, 1, 3, 1,
       2, 0, 3, 0, 1, 0, 1, 2, 3, 0, 2, 3])

In [87]:
women_kids_changed_sample_mean = np.mean(women_kids_changed_sample,dtype='float')

In [88]:
women_kids_changed_sample_mean

1.55

In [89]:
women_kids_changed_sample_stddev = np.std(women_kids_changed_sample,ddof=0,dtype='float')

In [90]:
women_kids_changed_sample_stddev

1.116915395184434

In [91]:
women_kids_changed_sample_results = s_tests.ztest(women_kids_changed_sample, value=1.48, alternative='two-sided', ddof=0)

In [92]:
women_kids_changed_sample_results

(0.6267260734501837, 0.5308388096316876)

In [93]:
hyp_test(women_kids_changed_sample,0.99,'lr')

Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


## **POPULATION_PROPORTION**
- **Using Z Score Chart/Table**
---

#### ***Question-7*** :: **As per the HR co-ordinators of different organizations in Germany, atleast 90% of the interns got permanent employment after completion of their internship. A researcher believes this number is too high. He gathered the interns hiring stats from 200 local companies. Test the claim at 90% and 99% LOC.**

In [94]:
interns_hired_sample = np.random.randint(low=10,high=16,size=200,dtype='int')

In [95]:
interns_hired_sample

array([13, 12, 13, 10, 15, 12, 15, 14, 15, 12, 14, 12, 13, 15, 12, 15, 10,
       15, 14, 13, 13, 14, 11, 13, 12, 11, 14, 13, 13, 12, 12, 13, 12, 15,
       13, 12, 10, 14, 11, 14, 10, 15, 12, 12, 13, 13, 14, 14, 11, 10, 14,
       12, 12, 15, 10, 13, 11, 12, 15, 14, 12, 10, 14, 11, 14, 15, 13, 14,
       14, 10, 15, 12, 10, 11, 10, 12, 11, 10, 11, 10, 13, 15, 13, 13, 13,
       14, 11, 12, 15, 12, 10, 10, 15, 15, 10, 11, 13, 15, 10, 14, 11, 11,
       15, 12, 13, 11, 14, 13, 11, 15, 14, 15, 11, 12, 15, 12, 11, 14, 13,
       11, 15, 10, 15, 11, 15, 13, 14, 11, 13, 15, 15, 13, 10, 15, 10, 15,
       10, 13, 14, 14, 13, 13, 15, 15, 13, 13, 14, 10, 14, 11, 14, 15, 15,
       15, 11, 11, 12, 14, 10, 14, 13, 14, 13, 10, 12, 15, 15, 10, 15, 11,
       14, 10, 14, 15, 15, 10, 11, 12, 10, 15, 13, 10, 12, 10, 12, 11, 15,
       12, 13, 11, 11, 12, 11, 15, 11, 10, 11, 15, 12, 10])

In [96]:
interns_convereted_as_permanent = interns_hired_sample - np.random.randint(low=1,high=4,size=200,dtype='int')

In [97]:
interns_convereted_as_permanent

array([12, 10, 12,  8, 12, 11, 13, 12, 14, 10, 12,  9, 12, 13, 10, 13,  8,
       14, 13, 10, 11, 13,  9, 11, 10,  8, 13, 12, 12,  9, 10, 12, 11, 14,
       11,  9,  8, 12,  8, 12,  9, 13, 10,  9, 11, 10, 11, 12,  8,  8, 12,
        9,  9, 12,  8, 10, 10, 10, 12, 13, 10,  7, 12,  9, 11, 13, 12, 13,
       11,  7, 12,  9,  9,  9,  9, 11, 10,  7, 10,  7, 12, 13, 11, 10, 12,
       12,  9, 10, 14, 10,  7,  7, 12, 13,  8,  9, 11, 12,  8, 12, 10,  9,
       14, 11, 11,  8, 11, 11,  9, 14, 12, 14, 10,  9, 13, 10,  9, 12, 10,
       10, 12,  8, 13,  8, 14, 12, 11,  8, 11, 13, 12, 10,  7, 12,  7, 12,
        9, 10, 11, 13, 11, 12, 13, 12, 11, 11, 11,  9, 13,  9, 13, 13, 12,
       12,  9,  9, 10, 12,  9, 12, 12, 12, 11,  8, 10, 13, 13,  9, 14,  9,
       12,  8, 11, 13, 12,  7,  9,  9,  9, 12, 11,  8, 11,  7,  9, 10, 14,
       11, 12,  9, 10,  9,  9, 12, 10,  7,  9, 12, 11,  8])

In [98]:
interns_sample_prop = interns_convereted_as_permanent/interns_hired_sample

In [99]:
interns_sample_prop

array([0.92307692, 0.83333333, 0.92307692, 0.8       , 0.8       ,
       0.91666667, 0.86666667, 0.85714286, 0.93333333, 0.83333333,
       0.85714286, 0.75      , 0.92307692, 0.86666667, 0.83333333,
       0.86666667, 0.8       , 0.93333333, 0.92857143, 0.76923077,
       0.84615385, 0.92857143, 0.81818182, 0.84615385, 0.83333333,
       0.72727273, 0.92857143, 0.92307692, 0.92307692, 0.75      ,
       0.83333333, 0.92307692, 0.91666667, 0.93333333, 0.84615385,
       0.75      , 0.8       , 0.85714286, 0.72727273, 0.85714286,
       0.9       , 0.86666667, 0.83333333, 0.75      , 0.84615385,
       0.76923077, 0.78571429, 0.85714286, 0.72727273, 0.8       ,
       0.85714286, 0.75      , 0.75      , 0.8       , 0.8       ,
       0.76923077, 0.90909091, 0.83333333, 0.8       , 0.92857143,
       0.83333333, 0.7       , 0.85714286, 0.81818182, 0.78571429,
       0.86666667, 0.92307692, 0.92857143, 0.78571429, 0.7       ,
       0.8       , 0.75      , 0.9       , 0.81818182, 0.9    

In [100]:
interns_sample_prop_mean = round(np.mean(interns_sample_prop,dtype='float'),4)

In [101]:
interns_sample_prop_mean

0.8369

In [102]:
interns_sample_prop_stddev = round(np.std(interns_sample_prop),4)

In [103]:
interns_sample_prop_stddev

0.0656

In [104]:
interns_hired_sample_results = sm.stats.proportions_ztest(interns_sample_prop_mean*200,nobs=200,value=0.90,alternative='smaller'
                                                          ,prop_var=0.90)

In [105]:
interns_hired_sample_results

(-2.9745625261914124, 0.0014670332670623557)

#### ***CASE - I :: With 90% Level of Confidence***

In [106]:
hyp_test(interns_hired_sample_results,0.90,'l')

Researcher claim is right. Thus, rejected the Null Hypothesis at 0.9 L.O.C and 0.1 L.O.S


#### ***CASE - II :: With 99% Level of Confidence***

In [107]:
hyp_test(interns_hired_sample_results,0.99,'l')

Researcher claim is right. Thus, rejected the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


## **CHI_SQUARE_TEST**
---

#### ***Question-8*** :: **A pencil manufacturer requires that the mass of their pencils have S.D that won't exceed 0.08 gms. An inspector thinks that the S.D is larger. He samples 30 pencils & finds they have a mean mass of 1.62 gms & S.D 0.0804 gms. Test the claim at 0.05 L.O.S.**

In [108]:
pencils_sample = np.random.randint(1,4,size=30,dtype='int')

In [109]:
pencils_sample

array([3, 2, 1, 2, 1, 3, 3, 1, 2, 1, 3, 2, 2, 3, 2, 3, 1, 2, 1, 3, 2, 3,
       2, 2, 3, 1, 2, 1, 1, 1])

In [113]:
def chi_square_one_pop(f_exp,loc,test_tail=(['l','r','lr']),ddof=False,sample_data=False,sample_stddev=False):
    """
    Description: This function performs the Chi-square test for one population.
    
    Input parameters: It accepts below parameters:
        1. f_exp : Expected Standard Deviation value or Standard Deviation value from population
        2. loc : Level of confidence. Used this parameter to calculate the critical value
        3. test_tail : This parameter represents which kind of test you want to perform
        4. ddof : Degree of freedom. Use this paramter if you want to provide the adhoc value of dof
        5. sample_data : One dimensional array containing the sample data
        6. sample_stddev : You can provide the standard deviation of the sample directly as an input to perform the chi-sqaure test
        
    Returns:
        - In case of 'l' means left-tail test, it returns,
            - test_statistic
            - left critical value
            - p_value
        
        - In case of 'r' means right-tail test, it returns,
            - test_statistic
            - right critical value
            - p_value

        - In case of 'lr' means both-tail test, it returns,
            - test_statistic
            - left critical value
            - right critical value
            - p-value based on left tail
            - p-value based on right tail
"""
    def cal_mean_var_std(input_array):
        """
        Description: This function calculates the mean, variance and standard deviation of the 1 population.
        
        Input: It accepts below inp parameters:
            1. input_array : Sample Population-1
            
        Returns: Population-1:
                    - Mean
                    - Variance
                    - Standard Deviation
        """
        sample_data_mean = round(np.mean(input_array,dtype='float'),3)
        sample_data_stddev = round(np.mean(input_array,dtype='float'),3)
        sample_data_var = round(np.var(input_array,dtype='float'),3)
        return sample_data_mean, sample_data_var, sample_data_stddev
    
    if sample_stddev != False and sample_data == False and ddof != False:
        sample_data_stddev = sample_stddev
        sample_data_var = sample_data_stddev**2
        dof = ddof
    elif sample_data != False and sample_stddev == False and ddof == False:
        sample_data_mean, sample_data_stddev, sample_data_var = cal_mean_var_std(sample_data)
        total_obs = len(sample_data)
        dof = (total_obs-1)
    elif sample_data != False and sample_stddev == False and ddof != False:
        sample_data_mean, sample_data_stddev, sample_data_var = cal_mean_var_std(sample_data)
        dof = ddof
    
    f_exp_var = f_exp**2
    
    test_stat = (dof*sample_data_var)/f_exp_var
    
    def left_tail_crit_p_val(tail_test,c,df,test_statistic):
        """
        Description: This function is performing the left tail chi-square hypothesis testing.
        
        Input: It accepts below input parameters:
            1. tail_test : This should be 'l' as we are performing left-tail test
            2. c : Level of confidence
            3. df : Degree of freedom
            4. test_statistic : Test statistic that we have calculated from the sample data
        
        Returns: Left tail-test:
                    - Test Statistic
                    - Critical value
                    - p_value
        """
        alpha = 1 - c
        lower_tail_prob = alpha
        critical_val = scipy_stests.chi2.ppf(lower_tail_prob,dof)
        p_value = 1 - scipy_stests.chi2.cdf(test_stat,dof)
        return test_stat, critical_val, p_value
    
    def right_tail_crit_p_val(tail_test,c,df,test_statistic):
        """
        Description: This function is performing the right tail chi-square hypothesis testing.
        
        Input: It accepts below input parameters:
            1. tail_test : This should be 'r' as we are performing left-tail test
            2. c : Level of confidence
            3. df : Degree of freedom
            4. test_statistic : Test statistic that we have calculated from the sample data
        
        Returns: Right tail-test:
                    - Test Statistic
                    - Critical value
                    - p_value
        """
        lower_tail_prob = c
        critical_val = scipy_stests.chi2.ppf(lower_tail_prob,dof)
        p_value = 1 - scipy_stests.chi2.cdf(test_stat,dof)
        return test_stat, critical_val, p_value
         
    def two_tail_crit_p_val(tail_test,c,df,test_statistic):
        """
        Description: This function is performing the both or two tail chi-square hypothesis testing.
        
        Input: It accepts below input parameters:
            1. tail_test : This should be 'lr' as we are performing left-tail test
            2. c : Level of confidence
            3. df : Degree of freedom
            4. test_statistic : Test statistic that we have calculated from the sample data
        
        Returns: Both or Two tail-test:
                    - Test Statistic
                    - Left critical value
                    - Right critical value
                    - p value
        """  
        l_alpha_by_2 = (1 - c)/2
        l_lower_tail_prob = l_alpha_by_2
        l_critical_val = scipy_tests.chi2.ppf(l_lower_tail_prob,dof)
        r_lower_tail_prob = (1 - l_alpha_by_2)
        r_critical_val = scipy_tests.chi2.ppf(r_lower_tail_prob,dof)
        p_value = 1 - scipy_tests.chi2.cdf(test_stat,dof)
        return test_stat, l_critical_val, r_critical_val, p_value
    
    if test_tail == 'l':
        test_stat, l_cri_val, p_value = left_tail_crit_p_val(tail_test=test_tail,c=loc,df=dof,test_statistic=test_stat)
        return test_stat, l_cri_val, p_value
    if test_tail == 'r':
        test_stat, r_cri_val, p_value = right_tail_crit_p_val(tail_test=test_tail,c=loc,df=dof,test_statistic=test_stat)
        return test_stat, r_cri_val, p_value
    if test_tail == 'lr':    
        test_stat, l_cric_val, r_cric_val, p_value = two_tail_crit_p_val(tail_test=test_tail,c=loc,df=dof,test_statistic=test_stat)
        return test_stat, l_cric_val, r_cric_val, p_value

In [114]:
chi_square_one_pop_results = chi_square_one_pop(f_exp=4,loc=0.99,ddof=39,test_tail='lr',sample_stddev=3.2)

In [115]:
chi_square_one_pop_results

(24.960000000000004, 19.99586787495632, 65.47557090346805, 0.9604831078918762)

In [116]:
def chi_square_hyp_test(test_stats_result,loc,kind_of_test=['l','r','lr']):
    alpha = 1 - loc
    if (kind_of_test == 'l') & (test_stats_result[-1] < alpha):
        print("Researcher claim is right. Thus, rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
    elif (kind_of_test == 'r') & (test_stats_result[-1] < alpha):
        print("Researcher claim is right. Thus, rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
    elif (kind_of_test == 'lr') & (test_stats_result[-1] < alpha) & ((test_stats_result[0] < test_stats_result[1]) or (test_stats_result[0]) > test_stats_result[2]):    
        print("Researcher claim is right. Thus, rejected the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))
    else:
        print("Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at {} L.O.C and {} L.O.S".format(loc,round(alpha,2)))

In [117]:
chi_square_hyp_test(chi_square_one_pop_results,0.99,'lr')

Researcher claim is wrong. Thus, fail to reject the Null Hypothesis at 0.99 L.O.C and 0.01 L.O.S


---
# **``Two_Populations``**
---

Import handmade image here

## **Large_Independent_Samples**
---
- #### **Independent samples means they are the true images of 2 populations which are:**
    - not related to each other or 
    - not having any relationship among them or
    - not dependent on each other or
    - both the people or items in the populations are different
- #### **This means n > 30, here n is sample size**

#### ***Question-9*** :: **A researcher thinks that grocery shoppers spend more when they haven't eaten. To test, he samples 41 shoppers who didn't eat breakfast. These people spent on average USD 72.27 with a standard deviation of USD 8.05. 52 shoppers who did eat breakfast spent on average USD 69.43 with standard deviation USD 9.22. Test the claim at 95% level of confidence.**

In [118]:
with_brkfst_sample1 = np.random.randint(low=60,high=90,size=52,dtype=np.int)

In [119]:
with_brkfst_sample1_mean = round(np.mean(with_brkfst_sample1),3)

In [120]:
with_brkfst_sample1_mean

75.25

In [121]:
with_brkfst_sample1_stddev = round(np.std(with_brkfst_sample1),3)

In [122]:
with_brkfst_sample1_stddev

8.886

In [123]:
with_brkfst_sample2 = np.random.randint(low=69,high=91,size=41,dtype=np.int)

In [124]:
with_brkfst_sample2_mean = round(np.mean(with_brkfst_sample2),3)

In [125]:
with_brkfst_sample2_mean

79.902

In [126]:
with_brkfst_sample2_stddev = round(np.std(with_brkfst_sample2),3)

In [127]:
with_brkfst_sample2_stddev

5.954

### **``Pooled_Large_Independent_Samples``**
#### **Ztest :: Function of Statsmodels.api.stats.ztest & Statsmodels.stats.weightstats.ztest packages only performs ``Pooled`` independent samples hypothesis testing.**  

In [128]:
two_pop_lrg_indp_samp_hyp_test_results = sm.stats.ztest(with_brkfst_sample1,with_brkfst_sample2,alternative='smaller',
                                                        value=0,usevar='pooled',ddof=0)

In [129]:
two_pop_lrg_indp_samp_hyp_test_results

(-2.8811236712622943, 0.0019813006236855347)

In [130]:
hyp_test(two_pop_lrg_indp_samp_hyp_test_results,0.95,'l')

Researcher claim is right. Thus, rejected the Null Hypothesis at 0.95 L.O.C and 0.05 L.O.S


#### **Manually calculating the Z Test Statistic for ``Large Independent Samples``**  

In [131]:
(with_brkfst_sample1_mean - with_brkfst_sample2_mean)/(np.sqrt(((with_brkfst_sample1_stddev**2)/52)+((with_brkfst_sample2_stddev**2)/41)))

-3.0134706282715853

##### **The above formula used for manual calculation of Z Test Statistic actually represents the formula for ``Not Pooled Sample`` large independent samples.**

### **``Not-Pooled_Large_Independent_Samples``**
#### **Ztest ::**
- **As function of Statsmodels.api.stats.ztest & Statsmodels.stats.weightstats.ztest packages only performs ``Pooled`` independent samples hypothesis testing. Therefore, write below function:**  

In [132]:
def ztest_notpooled(x1, x2=None, value=0, alternative='two-sided', usevar='pooled',
          ddof=1.):
    '''test for mean based on normal distribution, one or two samples

    In the case of two samples, the samples are assumed to be independent.

    Parameters
    ----------
    x1 : array_like, 1-D or 2-D
        first of the two independent samples
    x2 : array_like, 1-D or 2-D
        second of the two independent samples
    value : float
        In the one sample case, value is the mean of x1 under the Null
        hypothesis.
        In the two sample case, value is the difference between mean of x1 and
        mean of x2 under the Null hypothesis. The test statistic is
        `x1_mean - x2_mean - value`.
    alternative : string
        The alternative hypothesis, H1, has to be one of the following

           'two-sided': H1: difference in means not equal to value (default)
           'larger' :   H1: difference in means larger than value
           'smaller' :  H1: difference in means smaller than value

    usevar : string, 'unequal'
        ``unqual``, means the standard deviation of the samples is assumed to be
        not the same.
    ddof : int
        Degrees of freedom use in the calculation of the variance of the mean
        estimate. In the case of comparing means this is one, however it can
        be adjusted for testing other statistics (proportion, correlation)

    Returns
    -------
    tstat : float
        test statisic
    pvalue : float
        pvalue of the t-test
    '''
    import numpy as np
    from scipy import stats
    
    def _zstat_generic(value1, value2, std_diff, alternative, diff=0):
        '''generic (normal) z-test to save typing
        can be used as ztest based on summary statistics
        '''
        zstat = (value1 - value2 - diff) / std_diff
        if alternative in ['two-sided', '2-sided', '2s']:
            pvalue = stats.norm.sf(np.abs(zstat))*2
        elif alternative in ['larger', 'l']:
            pvalue = stats.norm.sf(zstat)
        elif alternative in ['smaller', 's']:
            pvalue = stats.norm.cdf(zstat)
        else:
            raise ValueError('invalid alternative')
        return zstat, pvalue
    
    if usevar == 'unequal':
        print("You are using Two populations whose variances are assumed to be ``Unequal`` or ``Not-pooled``")
#         raise NotImplementedError('only usevar="pooled" is implemented')

    x1 = np.asarray(x1)
    nobs1 = x1.shape[0]
    x1_mean = x1.mean(0)
    x1_var = x1.var(0)
    if x2 is not None:
        x2 = np.asarray(x2)
        nobs2 = x2.shape[0]
        x2_mean = x2.mean(0)
        x2_var = x2.var(0)
        var_pooled = ((x1_var/nobs1) + (x2_var/nobs2))
    else:
        var_pooled = x1_var / (nobs1 - ddof)
        x2_mean = 0

    std_diff = np.sqrt(var_pooled)
    #stat = x1_mean - x2_mean - value
    z_stat, pvalue = _zstat_generic(x1_mean, x2_mean, std_diff, alternative, diff=value)
    return z_stat, pvalue

In [133]:
z_test_not_pooled_results = ztest_notpooled(with_brkfst_sample1,with_brkfst_sample2,value=0,alternative='smaller',usevar='unequal')

You are using Two populations whose variances are assumed to be ``Unequal`` or ``Not-pooled``


In [134]:
z_test_not_pooled_results

(-3.0137485802222717, 0.0012902077547183167)

In [135]:
hyp_test(z_test_not_pooled_results,loc==0.95,'l')

Researcher claim is right. Thus, rejected the Null Hypothesis at False L.O.C and 1 L.O.S


## **Small_Independent_Samples**
---
- #### **Independent samples means they are the true images of 2 populations which are:**
    - not related to each other or 
    - not having any relationship among them or
    - not dependent on each other or
    - both the people or items in the populations are different
- #### **This means n < 30, here n is sample size**

#### ***Question-10*** :: **It is claimed that people who go to home improvement classes finish projects in less time. Group A is 10 people who attend a class. And, on average, they finish projects in 14.1 hrs with S.D of 2.3 hrs. Group B is 10 people who didn't attend a class and finish projects on average in 15 hrs with S.D of 2,4 hrs. Test the claim at 0.01 L.O.S.**

#### ***Question-11*** :: **A test preparation company claims that their class increases test scores. In city A, 15 students took the class and got an average score of 942 points with a S.D 103 points. In city B, 18 students did not take the class and got an average of 898 score with S.D of 95 points. Test the claim at 0.95 LOS. Because students are in different cities and assume variances not be equal.**

### **To be studied**

sm.stats.DescrStatsW()

1 + np.random.randn(2,4)

Compare Means